### ISSUE TITLE
Custom Trained NER Model Recognizing Entities it hasn't been trained for

### ISSUE THREAD

**Issue by spiralswimmer** [explosion/spaCy#3251]
I trained a customer ner model for recognising products from the retail space, the model was trained on using a blank. 

My training data only contained labels for entities I was interested in, there were no other labels for any of the other possible entities. There was only one label i.e. `PRODUCT`

Code used to compare results of entity tagging post training: 

```
nlp_en = spacy.load('en')

nlp_v1 = spacy.load('./test_v1')

sent = 'Have you ever been to Indonesia?'

doc_en = nlp_en(sent)

doc_v1 = nlp_v1(sent)

for ent in doc_en.ents:
    print(ent.text, ent.label_)
    
for ent in doc_v1.ents:
    print(ent.text, ent.label_)
```

Post training it seems to be working fine to recognise the entities however it is doing two odd things: 

1. It is recognising entities that it wasn't trained for such as city or individual names. The tags for these entities however are incorrectly marked as `PRODUCT`. For above code example I get the following result: 
```
Indonesia GPE
Indonesia PRODUCT
```

2. How does casing effect the NER?
For **'Have you ever been to givenchy?'** it gives the entity as **?** but for 
**'Have you ever been to Givenchy?'** it gives the entity correctly as **Givenchy**

## Your Environment
* **spaCy version:** 2.0.18
* **Platform:** Darwin-18.2.0-x86_64-i386-64bit
* **Python version:** 3.6.5
* **Models:** en

**Comment by ines** [comment]
Could you share more details on how you trained the model? How did you set up the blank base model and the labels?

**Comment by spiralswimmer** [comment]
I generated training phrases using the following code: 

```
training_phrases = '''I am looking for %s
Do we have %s at our store
Can I get %s?
Do we have any product from %s
Do we sell stuff from %s
Can you show me something from %s
Where can I find %s's cream'
Are you still out of %s's lotion?'
What is the difference between this and %s?
When can you deliver shower gel from %s?'''

training_phrases = training_phrases.split("\n")
product_list = ['Tata Harper', 'Velour Lashes', 'Verb', 'Violet Voss', 'Viseart']
TRAIN_DATA = []
for phrase in training_phrases:
    for product in product_list:
        text = phrase % product
        start_index = text.index(product)
        end_index = start_index + len(product)
        TRAIN_DATA.append((
          text, {"entities": [(start_index, end_index, 'PRODUCT')]}
        ))
```

In summary: 10 sentences with a product name embedded within them. This done for a 100+ products. Resulting in 100x10 training data points. 

For blank base model I used: 
```
nlp = spacy.blank("en")  # create blank Language class


if "ner" not in nlp.pipe_names:
        ner = nlp.create_pipe("ner")
        nlp.add_pipe(ner, last=True)
    # otherwise, get it so we can add label
else:
        ner = nlp.get_pipe("ner")
```

**Comment by ines** [comment]
Sorry, I think I misread your question. I thought your new model was actually predicting `GPE`, which would have been very strange.

I think what's happening here is that your synthetic training data is really one-sided and only includes examples of things that *are* products. So for all you know, your model could have learned something like "capitalised words are products" or "capitalised words in sentences with 'have' are products". So if you want to your model to be able to predict products, it also needs to see examples of what's **not** a product.

Try adding more examples of (reasonable!) sentences containing no products and add them to you data with `{"entities": []}`. And if that looks promising, create more examples, because 1000 is still very low for a new category. You might also want to try and find a raw corpus you can use to extract more diverse examples from *real* text to make your model more robust.

**Comment by spiralswimmer** [comment]
I am doing this for the first time so your response was really helpful. I am thinking of the following variations: 

1. Include sentences which don't have any products (as you suggested) 
2. Add products with different casings 
3. Should I add other entity types to my model as well?
4. Should I add examples such as: "Ratan Tata started a makeup company called Tata Harper" where I mark only "Tata Harper" as an entity?


Also any ideas around how to handle spelling mistakes in entity names, should I handle it at this layer or should that happen pre/post entity recognition?

**Comment by ines** [comment]
Yes, this sounds like a good plan. I still think you might want to try and find some "real" text, just to make sure your data isn't too biased by what you happen to come up with. If you don't have raw text, the [Reddit comments corpus](https://files.pushshift.io/reddit/comments/) has *lots*, including talk about the types of products you're interested in. 

> 3. Should I add other entity types to my training set as well?

Do you want to predict other entities, too? If so, then yes. If you're updating the model with a sentence and something like `[{'entities': [(0, 10, 'PRODUCT')]}]`, you're basically saying "Offsets 0 to 10 is a product entity, all other tokens are not an entity".

> Also any ideas around how to handle spelling mistakes in entity names, should I handle it at this layer or should that happen pre/post entity recognition?

Do you mean correcting spelling mistakes and resolving "Givancy" &rarr; "Givenchy"? In that case, yes, you want to do that in a separate step. You can predict things like that (although, not out-of-the-box with spaCy), or use a rule-based component and resolve them by calculating the [edit distance](https://en.wikipedia.org/wiki/Edit_distance). Once you've implemented that, you can add it to your spaCy entity spans as [custom extensiion attributes](https://spacy.io/usage/processing-pipelines#custom-components-attributes).
